### PR TITLE
[tune] Multi-objective support for Optuna

### LINKED ISSUE #18027: [tune] Multi-Objective Samplers Like NSGA-II found in Optuna
I'm using Optuna's NSGA-II multi-objective sampler in order to optimize over multiple performance metrics. However, one of the libraries I'm using is single-threaded; so, when I run the analysis with just Optuna, only 1 thread of my CPU is utilized. In an effort to maximize my CPU usage, as well as allow the possibility to expand the analysis to multiple servers I decided to use Ray. Awesome library by the way! Unfortunately, when looking through Ray Tune I didn't see any multi-objective samplers. So, I tried both to use Optuna with Ray by following Optuna's example, https://github.com/optuna/optuna-examples/blob/main/ray/ray_joblib.py, and Ray Tune's Optuna integration by following your example, https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/optuna_example.py.

When following Optuna's Ray example, only 1 thread was used and all workers were idle. When following your example with the code below, I received the following error, `ValueError: Invalid return or yield value. Either return/yield a single number or a dictionary object in your trainable function.` and if I make the objective function return a dictionary of objectives, I get a KeyError.

```
def objective(config):
    ...
    return (objective_1, objective_2, objective_3, objective_4, objective_5)

def run_optuna_tune():
    algo = OptunaSearch(
        mode=[
            "maximize",
            "maximize",
            "maximize",
            "minimize",
            "minimize",
        ],
        sampler=optuna.samplers.NSGAIISampler(),
    )
    algo = ConcurrencyLimiter(algo, max_concurrent=4)
    config = ...
    analysis = tune.run(
        objective,
        config=config,
        search_alg=algo,
        num_samples=N_TRIALS,
    )

if __name__ == "__main__":
    # Get a URL for the Ray dashboard.
    try:
        ray.init(address="auto")
    except ConnectionError:
        ray.init()
    # Disable the warning to suppress the log.
    # TODO(HideakiImamura): enable warning after https://github.com/ray-project/ray/pull/13865 is
    # released.
    ray.init(logging_level=logging.ERROR, ignore_reinit_error=True)

    run_optuna_tune()
```
Am I doing something wrong? Is there a better way to achieve my goals? Does Ray Tune have any Multi-objective sampler like NSGA-II? If no, no, and Ray Tune doesn't have any Multi-objective sampler like NSGA-II, can Ray Tune's integration of Optuna be changed to allow them and/or can Multi-objective samplers be added to Ray Tune?

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Adds support for multi-objective optimization to `OptunaSearch`. Adds test, example, docs.

## Related issue number

Closes https://github.com/ray-project/ray/issues/18027

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/suggest/optuna.py [hunk 1]
```diff
@@ -261,33 +293,56 @@ def __init__(self,
...
 
         pruner = ot.pruners.NopPruner()
         storage = ot.storages.InMemoryStorage()
 
+        if self._sampler:
+            sampler = self._sampler
+        elif isinstance(mode, list):
+            sampler = ot.samplers.MOTPESampler(seed=self._seed)
+        else:
+            sampler = ot.samplers.TPESampler(seed=self._seed)
```
Comment (himkt) [inline python/ray/tune/suggest/optuna.py:328]:
```suggestion
    else
            sampler = ot.samplers.TPESampler(seed=self._seed)
```

`MOTPESampler` is now deprecated and `TPESampler` works even if the objective has multiple values. 😸 

https://optuna.readthedocs.io/en/stable/reference/generated/optuna.samplers.MOTPESampler.html#optuna.samplers.MOTPESampler

File: python/ray/tune/suggest/optuna.py [hunk 2]
```diff
@@ -261,33 +293,56 @@ def __init__(self,
...
+                    "If ``mode`` is a list (multi-objective optimization "
+                    "case), ``metric`` must be defined.")
             # If only a mode was passed, use anonymous metric
             self._metric = DEFAULT_METRIC
 
         pruner = ot.pruners.NopPruner()
         storage = ot.storages.InMemoryStorage()
 
+        if self._sampler:
+            sampler = self._sampler
```
Comment (himkt) [inline python/ray/tune/suggest/optuna.py:324]: [note] Following samplers don't support multi-objective optimization. So, when `self._mode` is a list and `self._sampler` is one of them, Optuna would raise an exception (You can choose other samplers listed in https://optuna.readthedocs.io/en/stable/reference/samplers.html).

- https://optuna.readthedocs.io/en/stable/reference/generated/optuna.samplers.CmaEsSampler.html#optuna.samplers.CmaEsSampler
- https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.PyCmaSampler.html#optuna.integration.PyCmaSampler
- https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.SkoptSampler.html#optuna.integration.SkoptSampler

Comment (Yard1) [reply python/ray/tune/suggest/optuna.py:324]: I think it's fine to let Optuna raise it here. 

BTW. would it be possible for Optuna to add a property to its samplers that would inform whether the sampler can support multi-objective optimization?

File: python/ray/tune/suggest/optuna.py (hunk 1 above)
Comment (Yard1) [reply python/ray/tune/suggest/optuna.py:328]: Oh, that's great! I will leave MOTPE as we want to support older versions of Optuna too, but use TPE with the newer versions.

**Review by krfricke — APPROVED** [review]
Looks great! Thanks!

**Pull request merged**